### ISSUE TITLE
Invalid value in patient mismatch code field in audit spool file causes StringIndexOutOfBoundsException

### ISSUE THREAD

**Issue by Syrlander** [dcm4che/dcm4chee-arc-light#2704]
**Describe the bug**
After upgrading an installation of the mysql-secured-ui from version 5.22.0 to 5.22.3, I've moved all the audit logging files from the old ```Audit_Logger``` into the new ```Audit Logger``` path under ```$WILDFLY_HOME/standalone/data/audit-spool/```, as mentioned in the 5.22.2 release notes https://github.com/dcm4che/dcm4chee-arc-light/releases/tag/5.22.2. Doing so seems to have caused an error somewhere in the audit logger, as the server log files under ```$WILDFLY_HOME/standalone/log/server.log``` displays the error message:
```
2020-08-07 08:03:00,893 WARN  [org.dcm4chee.arc.audit.AuditService] (EE-ManagedScheduledExecutorService-default-Thread-2) Failed to mark [AuditSpoolFile=/opt/wildfly/standalone/data/audit-spool/Audit Logger/STORE_CREA-SCP1-SCP2-1.3.51.0.1.1.10.143.20.159.14882568.7942376] of [AuditLogger=Audit Logger] as failed.
: java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.substring(String.java:1967)
	at org.dcm4chee.arc.audit.EventID$Event.patMismatchEventTypeCode(EventID.java:150)
	at org.dcm4chee.arc.audit.EventID$Event.<init>(EventID.java:124)
	at org.dcm4chee.arc.audit.EventID.toEventIdentification(EventID.java:65)
	at org.dcm4chee.arc.audit.AuditService.auditStoreOrWADORetrieve(AuditService.java:598)
	at org.dcm4chee.arc.audit.AuditService.aggregateAuditMessage(AuditService.java:127)
	at org.dcm4chee.arc.audit.AuditService.auditAndProcessFile(AuditService.java:381)
	at sun.reflect.GeneratedMethodAccessor114.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.weld.bean.proxy.AbstractBeanInstance.invoke(AbstractBeanInstance.java:38)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:106)
	at org.dcm4chee.arc.audit.AuditService$Proxy$_$$_WeldClientProxy.auditAndProcessFile(Unknown Source)
	at org.dcm4chee.arc.audit.AuditScheduler.execute(AuditScheduler.java:116)
	at org.dcm4chee.arc.Scheduler.run(Scheduler.java:78)
	at org.jboss.as.ee.concurrent.ControlPointUtils$ControlledScheduledRunnable.run(ControlPointUtils.java:158)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at org.glassfish.enterprise.concurrent.internal.ManagedScheduledThreadPoolExecutor$ManagedScheduledFutureTask.access$201(ManagedScheduledThreadPoolExecutor.java:360)
	at org.glassfish.enterprise.concurrent.internal.ManagedScheduledThreadPoolExecutor$ManagedScheduledFutureTask.run(ManagedScheduledThreadPoolExecutor.java:511)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at org.glassfish.enterprise.concurrent.ManagedThreadFactoryImpl$ManagedThread.run(ManagedThreadFactoryImpl.java:227)
```
The first line of the error message points to the correct directory for audit spools (```AuditSpoolFile=/opt/wildfly/standalone/data/audit-spool/Audit Logger/```), so I don't know what's causing the issue. These errors happen quite frequently, thus creating huge log files and currently takes up approx. 90% of the server log files under ```$WILDFLY_HOME/standalone/log/```.

**Desktop:**
 - OS: CentOS 7
 - Version: 5.22.3

**Comment by vrindanayak** [comment]
@Syrlander : Can you share a sample DICOM object (on storing it to archive) which created this spool file?

**Comment by Syrlander** [comment]
@vrindanayak, I've added a sample DICOM object which produces the error once sent to the archive.
[sample-obj.zip](https://github.com/dcm4che/dcm4chee-arc-light/files/5040280/sample-obj.zip)

After sending the file using export from another installation, I get the following in the log file:
```
2020-08-07 10:40:32,771 INFO  [org.dcm4che3.net.Connection] (EE-ManagedExecutorService-default-Thread-2) Accept connection Socket[addr=/172.16.186.24,port=52179,localport=104]
2020-08-07 10:40:32,773 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572) >> A-ASSOCIATE-RQ
2020-08-07 10:40:32,774 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572) << A-ASSOCIATE-AC
2020-08-07 10:40:32,783 INFO  [org.dcm4che3.net.Dimse] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572) >> 1:C-STORE-RQ[pcid=1, prior=0
  cuid=1.2.840.10008.5.1.4.1.1.7 - Secondary Capture Image Storage
  iuid=1.3.753634535248669695546182077133271421180214039333443206970778 - ?
  tsuid=1.2.840.10008.1.2 - Implicit VR Little Endian]
2020-08-07 10:40:32,887 INFO  [org.dcm4chee.arc.patient.impl.PatientServiceEJB] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572): Create Patient[pk=635384, id=PatientID[pk=635384, id=test321test, issuer=null], name=Test^Person]
2020-08-07 10:40:32,893 INFO  [org.dcm4chee.arc.store.impl.StoreServiceEJB] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572): Create Study[pk=1105414, uid=1.3.122771537166176968844329998868803573890604198119150912947829, id=321TEST]
2020-08-07 10:40:32,933 INFO  [org.dcm4chee.arc.store.impl.StoreServiceEJB] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572): Create Series[pk=2189218, uid=1.3.753634535248669695546182077133271421180214039333443206970778, no=1, mod=OT]
2020-08-07 10:40:32,935 INFO  [org.dcm4chee.arc.store.impl.StoreServiceEJB] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572): Create Instance[pk=232566571, uid=1.3.753634535248669695546182077133271421180214039333443206970778, class=1.2.840.10008.5.1.4.1.1.7, no=1]
2020-08-07 10:40:32,936 INFO  [org.dcm4chee.arc.store.impl.StoreServiceEJB] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572): Create Location[pk=231842558, systemID=raid, path=2020/07/21/2ACB999A/83749B82/83749B82, tsuid=1.2.840.10008.1.2, size=1030, status=OK, objectType=DICOM_FILE]
2020-08-07 10:40:32,942 INFO  [org.dcm4chee.arc.store.impl.StoreServiceImpl] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572): Updated DB in 122 ms
2020-08-07 10:40:32,949 INFO  [org.dcm4chee.arc.export.mgt.impl.ExportManagerEJB] (EE-ManagedExecutorService-default-Thread-15966) Create ExportTask[pk=357402, exporterID=CalculateQueryAttributes, studyUID=1.3.122771537166176968844329998868803573890604198119150912947829, seriesUID=*, objectUID=*]
2020-08-07 10:40:32,951 INFO  [org.dcm4chee.arc.export.mgt.impl.ExportManagerEJB] (EE-ManagedExecutorService-default-Thread-15966) Create ExportTask[pk=357403, exporterID=SCP3ExportDescriptor, studyUID=1.3.122771537166176968844329998868803573890604198119150912947829, seriesUID=*, objectUID=*]
2020-08-07 10:40:32,954 INFO  [org.dcm4chee.arc.export.mgt.impl.ExportManagerEJB] (EE-ManagedExecutorService-default-Thread-15966) Create ExportTask[pk=357404, exporterID=CalculateStudySize, studyUID=1.3.122771537166176968844329998868803573890604198119150912947829, seriesUID=*, objectUID=*]
2020-08-07 10:40:32,955 INFO  [org.dcm4che3.net.Dimse] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572) << 1:C-STORE-RSP[pcid=1, status=0H
  cuid=1.2.840.10008.5.1.4.1.1.7 - Secondary Capture Image Storage
  iuid=1.3.753634535248669695546182077133271421180214039333443206970778 - ?
  tsuid=1.2.840.10008.1.2 - Implicit VR Little Endian]
2020-08-07 10:40:32,957 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572) >> A-RELEASE-RQ
2020-08-07 10:40:32,958 INFO  [org.dcm4che3.net.Association] (EE-ManagedExecutorService-default-Thread-15966) SCP2<-SCP1(6572) << A-RELEASE-RP
2020-08-07 10:40:33,008 INFO  [org.dcm4che3.net.Association] (EE-ManagedScheduledExecutorService-default-Thread-2) SCP2<-SCP1(6572): close Socket[addr=ssapc.regionh.top.local/172.16.186.24,port=52179,localport=104]
2020-08-07 10:40:39,239 INFO  [org.dcm4chee.arc.audit.AuditScheduler] (EE-ManagedScheduledExecutorService-default-Thread-1) start AuditScheduler.execute()
2020-08-07 10:40:39,241 WARN  [org.dcm4chee.arc.audit.AuditService] (EE-ManagedScheduledExecutorService-default-Thread-1) Failed to process [AuditSpoolFile=/opt/wildfly/standalone/data/audit-spool/Audit Logger/STORE_CREA-SCP3-SCP2-1.3.51.0.1.1.10.143.20.159.13910380.6911604] of [AuditLogger=Audit Logger].
: java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.substring(String.java:1967)
	at org.dcm4chee.arc.audit.EventID$Event.patMismatchEventTypeCode(EventID.java:150)
	at org.dcm4chee.arc.audit.EventID$Event.<init>(EventID.java:124)
	at org.dcm4chee.arc.audit.EventID.toEventIdentification(EventID.java:65)
	at org.dcm4chee.arc.audit.AuditService.auditStoreOrWADORetrieve(AuditService.java:598)
	at org.dcm4chee.arc.audit.AuditService.aggregateAuditMessage(AuditService.java:127)
	at org.dcm4chee.arc.audit.AuditService.auditAndProcessFile(AuditService.java:381)
	at sun.reflect.GeneratedMethodAccessor114.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.weld.bean.proxy.AbstractBeanInstance.invoke(AbstractBeanInstance.java:38)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:106)
	at org.dcm4chee.arc.audit.AuditService$Proxy$_$$_WeldClientProxy.auditAndProcessFile(Unknown Source)
	at org.dcm4chee.arc.audit.AuditScheduler.execute(AuditScheduler.java:116)
	at org.dcm4chee.arc.Scheduler.run(Scheduler.java:78)
	at org.jboss.as.ee.concurrent.ControlPointUtils$ControlledScheduledRunnable.run(ControlPointUtils.java:158)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at org.glassfish.enterprise.concurrent.internal.ManagedScheduledThreadPoolExecutor$ManagedScheduledFutureTask.access$201(ManagedScheduledThreadPoolExecutor.java:360)
	at org.glassfish.enterprise.concurrent.internal.ManagedScheduledThreadPoolExecutor$ManagedScheduledFutureTask.run(ManagedScheduledThreadPoolExecutor.java:511)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at org.glassfish.enterprise.concurrent.ManagedThreadFactoryImpl$ManagedThread.run(ManagedThreadFactoryImpl.java:227)
```

To add some context, there are three installations, which I've renamed SCP1, SCP2, SCP3 respectively (to keep the actual AETs hidden). The export I did to send the DICOM object to the archive was *from* SCP1 *to* SCP2 (the problem occurs on SCP2). On SCP2 there is an export rule setup, which exports all studies SCP2 receives to SCP3, if the study isn't from SCP3 itself.

I noticed in the above log snippet that the error message doesn't include SCP1, but only SCP2 and SCP3. So my thought is that the error is somehow related to the export rule, although under ```Configuration -> Devices -> dcm4chee-arc -> Extensions -> Device Extension -> Extensions -> Archive Device -> Child Objects``` I can't seem to be able to find any configurations related to audit logging for the ```Export Descriptor``` or ```Export Rule```.

**Comment by vrindanayak** [comment]
@Syrlander : The issue is not reproducible latest version also neither with [5.22.3 release](https://github.com/dcm4che/dcm4chee-arc-light/releases/tag/5.22.3) nor with [5.22.2 release](https://github.com/dcm4che/dcm4chee-arc-light/releases/tag/5.22.2). I've also updated the [5.22.3 release](https://github.com/dcm4che/dcm4chee-arc-light/releases/tag/5.22.3) as we missed updating a note regarding the refactoring of audit spool directories creation from audit logger names.

Can you also please share the spool file `STORE_CREA-SCP1-SCP2-1.3.51.0.1.1.10.143.20.159.14882568.7942376` causing this issue? It would be saved with `.failed` extension. 

This is the spool file that gets created [spool-file.zip](https://github.com/dcm4che/dcm4chee-arc-light/files/5040867/spool-file.zip) when I store your sample object to archive, and it is processed successfully (with all 3 versions)

**Comment by Syrlander** [comment]
@vrindanayak, the issue seems to be related to the exporter. I have a spool file (and the corresponding ```.failed``` file) for another study where the error occured:
[STORE_CREA-SCP3-SCP2-1.3.51.0.1.1.10.143.20.159.13817404.zip](https://github.com/dcm4che/dcm4chee-arc-light/files/5049758/STORE_CREA-SCP3-SCP2-1.3.51.0.1.1.10.143.20.159.13817404.zip)

As far as I can tell, the issue stems from an invalid configured patient mismatch code that gets passed to the constructor of the ```Code``` class, on ```line 151``` of dcm4chee-arc.audit.EventID (https://github.com/dcm4che/dcm4chee-arc-light/blob/master/dcm4chee-arc-audit/src/main/java/org/dcm4chee/arc/audit/EventID.java#L151)

Looking at the source for ```Code```, I can see that it performs an ```.indexOf(',')``` to determine the end index for trimming out the code value. I suspect that ```patMismatchCode```, passed to the constructor, doesn't contain any ```','``` which leads to the StringOutOfBoundsException during the ```trimsubstring``` operation.

If this issue isn't related to dcm4chee-arc, but rather an invalid study/configuration, do you know of any way to suppress these warnings? Since they take up a large amount of space in the log files.

**Comment by vrindanayak** [comment]
@Syrlander : As previously mentioned, field 30 in spool file gets populated **only if** patient mismatch occurred **during import of IMPAX reports**. I looked at the spool files you attached. These seem to be edited by someone, as they are invalid spool files, and are not created by the audit service code. The spool files contain only instance record entries but the first line (main line) which populates important information like calling / remote host names / users, study information, patient information etc is **missing** (See the spool file I attached in previous comment). Also indexing of each line starts with 0, field 30 in each of the lines has no value populated in the attached spool file.
eg. 
````
\\\\\\\\\\\\\\\\\1.2.840.10008.5.1.4.1.1.104.1\1.2.124.113532.80.22190.6098.20190429.135811.7159302\\\\\\\\0\\\0\\\\\\0\\\\
````

Also, as mentioned before, I've fixed the issue causing this exception and henceforth it will only be logged as a `LOG.info` message. I haven't been able to simulate the scenario which causes invalid entry in field 30 in first main line of spool file.
